src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,300 @@
+fn join_path(prefix: &str, volume_path: &str) -> Result<PathBuf> {
+    if volume_path.is_empty() {
+        return Err(anyhow!("volume path must not be empty"));
+    }
+    let b64_encoded_path = base64::encode(volume_path.as_bytes());
+
+    Ok(safe_path::scoped_join(prefix, b64_encoded_path)?)
+}
+
+// add writes the mount info (json string) of a direct volume into a filesystem path known to Kata Containers.
+pub fn add(volume_path: &str, mount_info: &str) -> Result<Option<String>> {
+    let mount_info_dir_path = join_path(KATA_DIRECT_VOLUME_ROOT_PATH, volume_path)?;
+
+    // create directory if missing
+    fs::create_dir_all(&mount_info_dir_path)?;
+
+    if !mount_info_dir_path.is_dir() {
+        return Err(anyhow!(
+            "{:?} is not a directory (metadata: {:?})",
+            mount_info_dir_path,
+            mount_info_dir_path.metadata()
+        ));
+    }
+
+    // This behavior of deserializing and serializing comes from
+    // https://github.com/kata-containers/kata-containers/blob/cd27ad144e1a111cb606015c5c9671431535e644/src/runtime/pkg/direct-volume/utils.go#L57-L79
+    // Assuming that this is for the purpose of validating the json schema.
+    let unserialized_mount_info: DirectVolumeMountInfo = serde_json::from_str(mount_info)?;
+
+    let mount_info_file_path = mount_info_dir_path.join(KATA_MOUNT_INFO_FILE_NAME);
+    let serialized_mount_info = serde_json::to_string(&unserialized_mount_info)?;
+    fs::write(mount_info_file_path, serialized_mount_info)?;
+
+    Ok(None)
+}
+
+// remove deletes the direct volume path including all the files inside it.
+pub fn remove(volume_path: &str) -> Result<Option<String>> {
+    let path = join_path(KATA_DIRECT_VOLUME_ROOT_PATH, volume_path)?;
+    // removes path and any children it contains.
+    fs::remove_dir_all(path)?;
+
+    Ok(None)
+}
+
+pub fn get_volume_mount_info(volume_path: &str) -> Result<DirectVolumeMountInfo> {
+    let mount_info_file_path =
+        join_path(KATA_DIRECT_VOLUME_ROOT_PATH, volume_path)?.join(KATA_MOUNT_INFO_FILE_NAME);
+    let mount_info_file = fs::read_to_string(mount_info_file_path)?;
+    let mount_info: DirectVolumeMountInfo = serde_json::from_str(&mount_info_file)?;
+
+    Ok(mount_info)
+}
+
+// get_sandbox_id_for_volume finds the id of the first sandbox found in the dir.
+// We expect a direct-assigned volume is associated with only a sandbox at a time.
+pub fn get_sandbox_id_for_volume(volume_path: &str) -> Result<String> {
+    let dir_path = join_path(KATA_DIRECT_VOLUME_ROOT_PATH, volume_path)?;
+    let paths = fs::read_dir(dir_path)?;
+    for path in paths {
+        let path = path?;
+        // compare with MOUNT_INFO_FILE_NAME
+        if path.file_name() == KATA_MOUNT_INFO_FILE_NAME {
+            continue;
+        }
+
+        let file_name = path.file_name();
+        // turn file_name into String and return it
+        let file_name = file_name.to_str().ok_or_else(|| {
```

Comment:
I think you could change this to, then you will be able to wrap the error with context.
```rust
let file_name = file_name.to_str().with_context(|| {
      format!(
          "failed to convert file_name {:?} to &str",
          file_name.to_string_lossy()
      )
  })?;
```